### PR TITLE
RFD: auto-generate the `tctl` resource reference

### PR DESCRIPTION
**Pull request by ptgott**

### CODE REVIEW

File: rfd/xxxx-config-resource-docs.md
Comment (codingllama) [inline rfd/xxxx-config-resource-docs.md:55]: +1 for the explicit mapping. It shouldn't be too much of a burden to keep.

File: rfd/xxxx-config-resource-docs.md
Comment (codingllama) [inline rfd/xxxx-config-resource-docs.md:86]: +1 for parsing Go sources too. It's pretty unlikely a resource won't have a Go representation, plus it doesn't tie us to necessarily to protos (or a particular proto generator).

File: rfd/xxxx-config-resource-docs.md
Comment (codingllama) [inline rfd/xxxx-config-resource-docs.md:313]: > However, these static analysis tools are intended for linting, not generating reference documentation

Design intention is key. It might work for a while, but it seems poor form to depend on it.

A program as simpler as we can make it, without unnecessary attachments, seems better long-term.

File: rfd/xxxx-config-resource-docs.md
Comment (codingllama) [inline rfd/xxxx-config-resource-docs.md:349]: Do we expect json tags to always exist?

File: rfd/xxxx-config-resource-docs.md
Comment (codingllama) [inline rfd/xxxx-config-resource-docs.md:363]: Could we make an educated guess and provide a way to override it? So simple numeric/string fields are taken from the Go type, as well as references to other structs?

Edit: I think that's what you explained below.

File: rfd/xxxx-config-resource-docs.md
Comment (codingllama) [inline rfd/xxxx-config-resource-docs.md:5]: Could you a required approvers section?

https://github.com/gravitational/teleport/blob/master/rfd/0000-rfds.md?plain=1#L106

**Review by codingllama — COMMENTED** [review]
Thanks for writing the RFD, Paul. Overall seems reasonable, although I confess I didn't spend too much time considering the implementation, as it seems alright for us to refine it as we go.

Would you mind following the RFD numbering/branching process described at https://github.com/gravitational/teleport/blob/master/rfd/0000-rfds.md?plain=1#L58-L78? I think we can keep this PR, since it already has some history, but maybe take the numbered branch too so it's clear it exists (and point it back to here?)

**Comment by codingllama** [comment]
Also, feel free to move out of draft, or alternatively let us know what's keeping you.

File: rfd/xxxx-config-resource-docs.md
Comment (ptgott) [reply rfd/xxxx-config-resource-docs.md:349]: Ah, that's a good point. Should the generator check whether the struct type has an `UnmarshalYAML` or `UnmarshalJSON` method before processing the struct declaration AST? 

If there is an `UnmarshalYAML` or `UnmarshalJSON` method, the generator would treat the struct as a custom type, since it won't be able to tell how struct field names translate into YAML field names. If there are no such methods, the generator would use the original struct field name.

File: rfd/xxxx-config-resource-docs.md
Comment (ptgott) [reply rfd/xxxx-config-resource-docs.md:330]: For now, I think the only requirement is that a GoDoc exists. I've suggested exiting with an error if there's no GoDoc above a field so we can add one. 

The only special rule is that, if a field name begins with the name of the field, remove the field name (and "is" if it exists) and capitalize the next first letter. This is to prevent field names that aren't user facing from making it into the docs, and also removes repetition with the first column of the table.

This GoDoc:

```go
// MyField is a field that specifies something.
```

...would become:

```
A field that specifies something
```

Otherwise, I think it's okay to move each field's GoDoc into the table without alteration.

File: rfd/xxxx-config-resource-docs.md
Comment (ptgott) [reply rfd/xxxx-config-resource-docs.md:363]: I'll add my response to this in my response to [this comment](https://github.com/gravitational/teleport/pull/25823#discussion_r1191510099).

File: rfd/xxxx-config-resource-docs.md
Comment (ptgott) [reply rfd/xxxx-config-resource-docs.md:366]: In this case, would it make sense to add an example YAML document for a custom type within the type's GoDoc?

E.g., for [`Labels`](https://github.com/gravitational/teleport/blob/c91da46765953688d218dcba4a4ce7acf606639f/api/types/role.go#L1137-L1140):

```go
// Labels is a wrapper around map
// that can marshal and unmarshal itself
// from scalar and list values
// Example YAML:
// 'env': 'test'
// '*': '*'
// 'region': ['us-west-1', -eu-central-1]
// 'reg': ^us-west-1|eu-central-1$'
// ---
type Labels map[string]utils.Strings
```

The generator could then exit with an error if a GoDoc for a custom type does not include an example YAML document, which it interprets as everything between `Example YAML` and the following `---`. If it can't parse the example YAML document as YAML, it also exits with an error.

File: rfd/xxxx-config-resource-docs.md
Comment (codingllama) [reply rfd/xxxx-config-resource-docs.md:357]: I'm more interested in what our "requirements" are - it's alright if we want all structs to have JSON tags, as long as we say so here.

Special handling of custom unmarshal methods seems like a good idea, or at the very least some warnings from the generator.

File: rfd/xxxx-config-resource-docs.md
Comment (codingllama) [reply rfd/xxxx-config-resource-docs.md:337]: > if a field name begins with the name of the field, remove the field name

The caveat I see here is that sometimes the field name is part of the sentence in a way that could have this break, but I suppose we can be careful and edit the types we care about.

File: rfd/xxxx-config-resource-docs.md
Comment (codingllama) [reply rfd/xxxx-config-resource-docs.md:374]: That sounds fine to me, +1.

File: rfd/xxxx-config-resource-docs.md
Comment (codingllama) [inline rfd/xxxx-config-resource-docs.md:566]: Assuming this is not already a part of our release procedure, +1.

Better yet if we can make it part of the release procedure, I think.

File: rfd/xxxx-config-resource-docs.md
Comment (codingllama) [inline rfd/xxxx-config-resource-docs.md:9]: Add a 2nd approver for another pair of eyes? Maybe @zmb3, since he's already reviewed.

**Review by codingllama — APPROVED** [review]
Thanks for putting this together, Paul.

No overriding concerns on my part.

**Comment by codingllama** [comment]
> Would you mind following the RFD numbering/branching process described at https://github.com/gravitational/teleport/blob/master/rfd/0000-rfds.md?plain=1#L58-L78? I think we can keep this PR, since it already has some history, but maybe take the numbered branch too so it's clear it exists (and point it back to here?)

Friendly ping on this? @ptgott

File: rfd/xxxx-config-resource-docs.md
Comment (ptgott) [reply rfd/xxxx-config-resource-docs.md:337]: > The caveat I see here is that sometimes the field name is part of the sentence in a way that could have this break, but I suppose we can be careful and edit the types we care about.

We could also simply replace all mentions of the original field name in the GoDoc for a field with mentions of the one included in the docs, i.e., the struct tag. I'll include that in the RFD.

File: rfd/xxxx-config-resource-docs.md
Comment (ptgott) [reply rfd/xxxx-config-resource-docs.md:14]: That's right, the scope of this RFD is for resources you can apply via `tctl`. The terminology mistake is also based on my own misconceptions, though I'm not sure the current docs are clear on this either.

It looks like we have three different uses of "resource" in the context of Teleport:
- Dynamic configuration resources (i.e., alternatives to static configuration files, as you mention)
- Teleport resources, e.g., `role`s and local `user`s
- Infrastructure resources you can enroll with Teleport, either via Teleport resources or configuration files

I think these three usages are okay as long as we make sure we're precise about them. I know this is off-topic, but @zmb3 should we mention these distinctions in the Core Concepts page?

File: rfd/xxxx-config-resource-docs.md
Comment (ptgott) [reply rfd/xxxx-config-resource-docs.md:62]: I suggested manually listing the resources because that's how we do it with the [Kubernetes operator](https://github.com/gravitational/teleport/blob/09adc3ea5cddac5c1fea479ddbfc0075c864677f/integrations/operator/crdgen/main.go#L97-L113). Since we need to develop controllers for the operator, this makes sense in that context, but you're right that it's not great for `tctl`.

I like the idea of including all resources with an embedded `Metadata` field.

Comment (ptgott) [reply rfd/xxxx-config-resource-docs.md:69]: @codingllama @zmb3 Come to think of it, are there any structs with a `github.com/gravitational/teleport/api/types.Metadata` field that we would want to exclude? Would it be possible to eliminate the configuration entirely?

Comment (codingllama) [reply rfd/xxxx-config-resource-docs.md]: Not sure, but that should be easy to search for. If there aren't any, +1 to automatically finding resources.

**Pull request closed without merging**
